carla/cf_models/catalog/dice/model.py

```diff
@@ -0,0 +1,87 @@
+import dice_ml
+import pandas as pd
+
+from ...api import CFModel
+
+
+class Dice(CFModel):
+    def __init__(self, mlmodel, data):
+        """
+        Constructor for Dice model
+        Implementation can be seen at https://github.com/interpretml/DiCE
+
+        Restrictions:
+        ------------
+        -   Only the model agnostic approach (backend: sklearn) is used in our implementation.
+        -   ML model needs to have a transformation pipeline for normalization, encoding and feature order.
+            See pipelining at carla/models/catalog/catalog.py for an example ML model class implementation
+
+        Parameters
+        ----------
+        mlmodel : models.api.MLModel
+            ML model to build counterfactuals for.
+        data : data.api.Data
+            Underlying dataset we want to build counterfactuals for.
+        """
+        # Prepare data for dice data structure
+        self.data = data
+        self._data_raw = data.raw.copy()
+        self._cont_feat = data.continous
+        self._target_name = data.target
+        self._dice_data = dice_ml.Data(
+            dataframe=self._data_raw,
+            continuous_features=self._cont_feat,
+            outcome_name=self._target_name,
+        )
+
+        # Build dice model structure
+        # Since our own model class MLModel resembles the sklearn structure it may suffice to use the sklearn
+        # backend implementation made by dice
+        self._backend = "sklearn"
+
+        self._mlmodel = mlmodel
+        self._dice_model = dice_ml.Model(model=mlmodel, backend=self._backend)
+
+        self._dice = dice_ml.Dice(self._dice_data, self._dice_model, method="random")
+
+    @property
+    def dice_model(self):
+        return self._dice
+
+    def get_counterfactuals(self, factuals, num_of_cf, desired_class):
+        """
+        Compute a certain number of counterfactuals per factual example.
+
+
+        Parameters
+        ----------
+        factuals : pd.DataFrame
+            DataFrame containing all samples for which we want to generate counterfactual examples.
+            All instances should belong to the same class.
+        num_of_cf : int
+            Number of counterfactuals we want to generate per factual
+        desired_class : int
+            The target class we want to reach for our factuals
+
+        Returns
+        -------
+
+        """
+
+        # Prepare factuals
+        querry_instances = factuals.copy()
```

Comment:
Is `querry_instances` a known term? Or should we rather name it factuals aswell? Also why do we need to `.copy()` do we expect that the consuming function is alerting the list and we need the original somewhere else in the flow?

Reply:
I used copy to prevent side-effects outside the method, this happened a lot in the last project.
Querry instances are used in the dice implementation and  I wanted to make it distinguishable to the parameter.

Reply:
Cool, naming makes sense then since the scope is limited to dice here